staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go

```diff
@@ -177,25 +199,153 @@ func generateKey(length int) (key []byte, err error) {
 	return key, nil
 }
 
+// NewKDFExtendedNonceGCMTransformerFromSeedUnsafe is the same as NewGCMTransformer but trades storage,
+// memory and CPU to work around the limitations of AES-GCM's 12 byte nonce size.  It is marked
+// as unsafe because it assumes that the input seed is a cryptographically strong key that is at least
+// 32 bytes in length.  Callers that do need to supply a specific seed must use
+// NewKDFExtendedNonceGCMTransformerWithUniqueSeed instead to guarantee that a secure seed is used.
+// Unlike NewGCMTransformer, this function is immune to the birthday attack because a new key is generated
+// per encryption via a key derivation function: KDF(seed, random_bytes) -> key.  The derived key is
+// only used once as an AES-GCM key with a random 12 byte nonce.  This avoids any concerns around
+// cryptographic wear out (by either number of encryptions or the amount of data being encrypted).
+// No specific rotation schedule is required for the seed.  NewKDFExtendedNonceGCMTransformerWithUniqueSeed
```

Comment:
@mikedanese do you agree with the statements around seed rotation:

> No specific rotation schedule is required for the seed.

> reusing the seed returned from this function is safe to do over time and across process restarts.

I am still working on confirming if MS crypto folks have an opinion on how often (if ever) the seed needs to be rotated (assuming the KEK is not being rotated / rotated very rarely - KEK rotation synthetically causes the seed to rotate).

Reply:
I would rephrase this as, "Speaking on the cryptographic safety, the limit on the number of operations that can be preformed with a single key with randomly generated nonces is not practically reachable. Thus, the scheme does not impose any specific requirements on the key rotation schedule."

Then say an organization standards or compliance policies around rotation can be implemented externally by rotating the kek.

Reply:
> I think the recommendation here is to rotate DEK/seed every month.

My reading of the MS crypto folks recommendations was that 'we could rotate the seed every month just to make sure it all works.'  I will circle back with them on this.  If need be we can update the docs next release (or update the code to do automatic monthly seed rotation).